src/databricks_ai_bridge/genie.py

```diff
@@ -59,15 +59,33 @@ def _parse_query_result(resp) -> Union[str, pd.DataFrame]:
 
         rows.append(row)
 
-    query_result = pd.DataFrame(rows, columns=header).to_markdown()
-
+    dataframe = pd.DataFrame(rows, columns=header)
+    query_result = dataframe.to_markdown()
     tokens_used = _count_tokens(query_result)
-    while tokens_used > MAX_TOKENS_OF_DATA:
-        rows.pop()
-        query_result = pd.DataFrame(rows, columns=header).to_markdown()
-        tokens_used = _count_tokens(query_result)
 
-    return query_result.strip() if query_result else query_result
+    # If the full result fits, return it
+    if tokens_used <= MAX_TOKENS_OF_DATA:
+        return query_result.strip()
+
+    # Binary search to find the maximum number of rows that fit within the token limit
```

Comment:
Maybe we can use https://docs.python.org/3/library/bisect.html#bisect.bisect_left to do this?